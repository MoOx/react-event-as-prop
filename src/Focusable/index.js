// @flow

import React, { Component } from "react"

type PropsType = {
  onFocus?: Function | boolean,
  onBlur?: Function | boolean,
}

type StateType = {
  focused: boolean,
}

// @todo try a more complex type for ComposedComponent + return type
export default (ComposedComponent: Function) => {
  class Focused extends Component<void, PropsType, StateType> {
    props: PropsType

    state: StateType = {
      focused: false,
    }

    handleFocus: Function = (event: SyntheticEvent): void => {
      this.setState({ focused: true })
      if (typeof this.props.onFocus === "function") {
        this.props.onFocus(event)
      }
    }

    handleBlur: Function = (event: SyntheticEvent): void => {
      this.setState({ focused: false })
      if (typeof this.props.onBlur === "function") {
        this.props.onBlur(event)
      }
    }

    render(): React$Element<any> {
      return (
        <ComposedComponent
          { ...this.props }
          { ...this.state }
          onFocus={ this.handleFocus }
          onBlur={ this.handleBlur }
        />
      )
    }
  }

  return Focused
}
