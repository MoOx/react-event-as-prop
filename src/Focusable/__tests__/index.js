import React from "react"
import renderer from "react-test-renderer"

import Focusable from ".."

const Compo = ({ focused, ...props }: { focused: boolean }) => (
  <span { ...props }>{ focused ? "focused" : "nope" }</span>
)
const Composed = Focusable(Compo)

test("focused props changes when focused", () => {
  const component = renderer.create(
    <Composed
      test
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  tree.props.onFocus()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  tree.props.onBlur()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

let flag = 0
const handleFocus = () => flag = 1
const handleBlur = () => flag = 2

test("focused props changes when focused and execute on{Event}", () => {
  const component = renderer.create(
    <Composed
      test
      onFocus={ handleFocus }
      onBlur={ handleBlur }
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(0)

  tree.props.onFocus()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(1)

  tree.props.onBlur()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(flag).toEqual(2)
})
