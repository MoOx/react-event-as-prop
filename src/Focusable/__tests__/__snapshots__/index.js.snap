exports[`test focused props changes when focused 1`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test focused props changes when focused 2`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
  test={true}>
  focused
</span>
`;

exports[`test focused props changes when focused 3`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test focused props changes when focused and execute on{Event} 1`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
  test={true}>
  nope
</span>
`;

exports[`test focused props changes when focused and execute on{Event} 2`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
  test={true}>
  focused
</span>
`;

exports[`test focused props changes when focused and execute on{Event} 3`] = `
<span
  onBlur={[Function]}
  onFocus={[Function]}
  test={true}>
  nope
</span>
`;
